model/src/main/kotlin/Provenance.kt

```diff
@@ -19,36 +19,69 @@
 
 package org.ossreviewtoolkit.model
 
-import com.fasterxml.jackson.annotation.JsonAlias
 import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import com.fasterxml.jackson.databind.deser.BeanDeserializer
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer
+import com.fasterxml.jackson.module.kotlin.treeToValue
 
 import java.time.Instant
 
 /**
- * Provenance information about the scanned source code. Either [sourceArtifact] or [vcsInfo] can be set to a non-null
- * value. If both are null this indicates that no provenance information is available.
+ * Provenance information about the origin of source code.
  */
-data class Provenance(
+@JsonDeserialize(using = ProvenanceDeserializer::class)
+sealed class Provenance {
     /**
-     * The time when the source code was downloaded, or [Instant.EPOCH] if unknown (e.g. for source code that was
-     * downloaded separately from running ORT).
+     * True if this [Provenance] refers to the same source code as [pkg], assuming that it belongs to the package id.
      */
-    @JsonAlias("downloadTime")
-    val downloadTime: Instant = Instant.EPOCH,
+    abstract fun matches(pkg: Package): Boolean
+
+    fun asArtifact(): ArtifactProvenance? = this as? ArtifactProvenance
+    fun asRepository(): RepositoryProvenance? = this as? RepositoryProvenance
+}
+
+object UnknownProvenance : Provenance() {
+    override fun matches(pkg: Package): Boolean = false // TODO: or true?
```

Comment:
Regarding the `TODO`, I believe it should stay at `false`. If it's unknown you can never say for sure it matches another provenance, and also `UnknownProvenance` should not match `UnknownProvenance`.